omnigibson/prims/rigid_prim.py

```diff
@@ -317,48 +340,87 @@ def set_position_orientation(self, position=None, orientation=None):
-        # Return cached pose if we're kinematic-only
-        if self.kinematic_only and self._kinematic_world_pose_cache is not None:
-            return self._kinematic_world_pose_cache
+    def get_position_orientation(self, frame: Literal["world", "scene", "parent"] = "world"):
+        """
+        Gets prim's pose with respect to the specified frame.
 
-        pos, ori = self._rigid_prim_view.get_world_poses()
+        Args:
+            frame (Literal): frame to get the pose with respect to. Default to world. parent frame
+            get position relative to the object parent. scene frame get position relative to the scene.
 
-        assert np.isclose(
-            np.linalg.norm(ori), 1, atol=1e-3
-        ), f"{self.prim_path} orientation {ori} is not a unit quaternion."
+        Returns:
+            2-tuple:
+                - 3-array: (x,y,z) position in the specified frame
+                - 4-array: (x,y,z,w) quaternion orientation in the specified frame
+        """
 
-        pos = pos[0]
-        ori = ori[0][[1, 2, 3, 0]]
-        if self.kinematic_only:
-            self._kinematic_world_pose_cache = (pos, ori)
-        return pos, ori
+        if frame == "world" or frame == "scene":
 
-    def set_local_pose(self, position=None, orientation=None):
-        # Invalidate kinematic-only object pose caches when new pose is set
-        if self.kinematic_only:
-            self.clear_kinematic_only_cache()
-        if position is not None:
-            position = np.asarray(position)[None, :]
-        if orientation is not None:
-            orientation = np.asarray(orientation)[None, [3, 0, 1, 2]]
-        self._rigid_prim_view.set_local_poses(position, orientation)
-        PoseAPI.invalidate()
+            if self.kinematic_only and self._kinematic_world_pose_cache is not None:
+                return self._kinematic_world_pose_cache
+
+            positions, orientations = self._rigid_prim_view.get_world_poses()
+
+            assert np.isclose(
+                np.linalg.norm(orientations), 1, atol=1e-3
+            ), f"{self.prim_path} orientation {orientations} is not a unit quaternion."
+
+            position = positions[0]
+            orientation = orientations[0][[1, 2, 3, 0]]
+
+            # Cache pose if we're kinematic-only
+            if self.kinematic_only:
+                self._kinematic_world_pose_cache = (position, orientation)
+
+        else:
+
+            # Return cached pose if we're kinematic-only
+            if self.kinematic_only and self._kinematic_local_pose_cache is not None:
+                return self._kinematic_local_pose_cache
+
+            positions, orientations = self._rigid_prim_view.get_local_poses()
+
+            position = positions[0]
+            orientation = orientations[0][[1, 2, 3, 0]]
+
+            if self.kinematic_only:
+                self._kinematic_local_pose_cache = (position, orientation)
+
+        # If we are in a scene, compute the scene-local transform
```

Comment:
same, let's try to not repeat these transformation code, put them into a util function